src/Coordination/KeeperStorage.cpp

```diff
@@ -198,356 +176,650 @@ void KeeperStorage::Node::removeChild(StringRef child_path)
+{
+    auto parent_path = parentPath(path);
+    auto node_it = container.find(parent_path);
+
+    if (node_it == container.end())
+        return false;
+
+    if (node_it->value.stat.ephemeralOwner != 0)
+        return false;
+
+    if (container.contains(path))
+        return false;
+
+    KeeperStorage::Node created_node;
+
+    uint64_t acl_id = acl_map.convertACLs(node_acls);
+    acl_map.addUsage(acl_id);
+
+    created_node.acl_id = acl_id;
+    created_node.stat = stat;
+    created_node.setData(std::move(data));
+    created_node.is_sequental = is_sequental;
+    auto [map_key, _] = container.insert(path, created_node);
+    /// Take child path from key owned by map.
+    auto child_path = getBaseName(map_key->getKey());
+    container.updateValue(parent_path, [child_path](KeeperStorage::Node & parent) { parent.addChild(child_path); });
+
+    if (is_ephemeral)
+        ephemerals[session_id].emplace(path);
+
+    return true;
+};
+
+bool KeeperStorage::removeNode(const std::string & path, int32_t version)
+{
+    auto node_it = container.find(path);
+    if (node_it == container.end())
+        return false;
+
+    if (version != -1 && version != node_it->value.stat.version)
+        return false;
+
+    if (node_it->value.stat.numChildren)
+        return false;
+
+    auto prev_node = node_it->value;
+    if (prev_node.stat.ephemeralOwner != 0)
+    {
+        auto ephemerals_it = ephemerals.find(prev_node.stat.ephemeralOwner);
+        ephemerals_it->second.erase(path);
+        if (ephemerals_it->second.empty())
+            ephemerals.erase(ephemerals_it);
+    }
+
+    acl_map.removeUsage(prev_node.acl_id);
+
+    container.updateValue(
+        parentPath(path),
+        [child_basename = getBaseName(node_it->key)](KeeperStorage::Node & parent) { parent.removeChild(child_basename); });
+
+    container.erase(path);
+    return true;
+}
+
 
 struct KeeperStorageRequestProcessor
 {
     Coordination::ZooKeeperRequestPtr zk_request;
 
-    explicit KeeperStorageRequestProcessor(const Coordination::ZooKeeperRequestPtr & zk_request_)
-        : zk_request(zk_request_)
-    {}
-    virtual std::pair<Coordination::ZooKeeperResponsePtr, Undo> process(KeeperStorage & storage, int64_t zxid, int64_t session_id, int64_t time) const = 0;
-    virtual KeeperStorage::ResponsesForSessions processWatches(KeeperStorage::Watches & /*watches*/, KeeperStorage::Watches & /*list_watches*/) const { return {}; }
-    virtual bool checkAuth(KeeperStorage & /*storage*/, int64_t /*session_id*/) const { return true; }
+    explicit KeeperStorageRequestProcessor(const Coordination::ZooKeeperRequestPtr & zk_request_) : zk_request(zk_request_) { }
+    virtual Coordination::ZooKeeperResponsePtr process(KeeperStorage & storage, int64_t zxid, int64_t session_id, int64_t time) const = 0;
+    virtual std::vector<KeeperStorage::Delta>
+    preprocess(KeeperStorage & /*storage*/, int64_t /*zxid*/, int64_t /*session_id*/, int64_t /*time*/) const
```

Comment:
As far as I understand it's strictly required to call `preprocess` before `process(Local)`. Maybe verify it in some way?

Reply:
I will add a lot of checks for things like that in #37252 because of the generated ZXID.